aten/src/ATen/native/cuda/group_norm_kernel.cu

```diff
@@ -614,22 +727,44 @@ void GroupNormKernelImplInternal(
     // using manual kernel here. Make it using gpu_kernel_multiple_outputs once
     // the issue fixed.
     const int64_t B = (N * C + kCUDANumThreads - 1) / kCUDANumThreads;
+
     ComputeFusedParamsCUDAKernel<T><<<B, kCUDANumThreads, 0, cuda_stream>>>(
         N, C, G, mean_data, rstd_data, gamma_data, beta_data, a_data, b_data);
     C10_CUDA_KERNEL_LAUNCH_CHECK();
 
-    auto iter = TensorIteratorConfig()
-                    .check_all_same_dtype(std::is_same<T, T_ACC>::value)
-                    .resize_outputs(false)
-                    .add_owned_output(Y.view({N * C, HxW}))
-                    .add_owned_const_input(X.view({N * C, HxW}))
-                    .add_owned_input(a.view({N * C, 1}))
-                    .add_owned_input(b.view({N * C, 1}))
-                    .build();
-    gpu_kernel(iter, [] GPU_LAMBDA(T x, T_ACC a, T_ACC b) -> T {
-      return a * static_cast<T_ACC>(x) + b;
-    });
+    switch (x_format) {
+      case MemoryFormat::Contiguous: {
+        TensorIterator iter =
+            TensorIteratorConfig()
+                .check_all_same_dtype(std::is_same<T, T_ACC>::value)
+                .resize_outputs(false)
+                .add_owned_output(Y.view({N * C, HxW}))
+                .add_owned_const_input(X.view({N * C, HxW}))
+                .add_owned_input(a.view({N * C, 1}))
+                .add_owned_input(b.view({N * C, 1}))
+                .build();
+        gpu_kernel(iter, [] GPU_LAMBDA(T x, T_ACC a, T_ACC b) -> T {
+          return a * static_cast<T_ACC>(x) + b;
+        });
+
+        break;
+      }
+      case MemoryFormat::ChannelsLast: {
+        int num_blocks =
+            (N * height * width * C + kCUDANumThreads - 1) / kCUDANumThreads;
+
+        ApplyScaleBiasNHWCKernel<T>
```

Comment:
Could the previous `TensorIterator` code just be adapted by adding a permute before the `view` here?

Reply:
Do you mean to use `permute` to convert them to NCHW, and then convert them back to NHWC after it's done? Please correct me if i'm wrong, I apologize as i'm not that faimliar with getting these iterators to work. 

Something like this?

```
auto X_permuted = X.permute({0, 3, 1, 2}); // N, C, H, W
auto Y_permuted = Y.permute({0, 3, 1, 2}); // N, C, H, W

TensorIterator iter =
    TensorIteratorConfig()
        .check_all_same_dtype(std::is_same<T, T_ACC>::value)
        .resize_outputs(false)
        .add_owned_output(Y_permuted.view({N * C, H * W}))
        .add_owned_const_input(X_permuted.view({N * C, H * W}))
        .add_owned_input(a.view({N * C, 1}))
        .add_owned_input(b.view({N * C, 1}))
        .build();
 gpu_kernel(iter, [] GPU_LAMBDA(T x, T_ACC a, T_ACC b) -> T {
  return a * static_cast<T_ACC>(x) + b;
});

Y = Y_permuted.permute({0, 2, 3, 1}); // N, H, W, C
```

Reply:
Yes, something like that---I'm not sure if `TensorIterator` is smart about getting the right memory access pattern for the permuted Tensor but it's worth an attempt

Reply:
@eqy So I get this with the above code

```RuntimeError: view size is not compatible with input tensor's size and stride (at least one dimension spans across two contiguous subspaces). Use .reshape(...) instead.
``` 

Am I missing some steps here or is it just not straight forward to work with these iterators when the tensors are in NHWC?

Reply:
Reshape should be fine in this case as well (same semantics as what we intend).

I wouldn't say it's really about compile time in this case. `TensorIterator` kernels save a lot of boilerplate code (imagine writing the same kernel every time for every pointwise OP variant but they also do a good job in achieving high B/W utilization for common-case workloads. There's a fair amount of optimization work that was done by e.g., @zasdfgbnm and if you look at the underlying kernels [here](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cuda/CUDALoops.cuh) they dispatch to there's the same optimizations that you would find in manually optimized kernels (vectorization, loop unrolling, etc.).

Reply:
Hm, I can't seem to get the above code to work with the `reshape` either - 100% mismatched elements in the tests lol. I've tried quite a few different approaches but I can't seem to get it right...

I can spend some time optimizing this kernel if need be with some vectorizing, but I'll have to defer to you on whether or not that level of hand-tuned optimizations are less maintainable. 

Do you think you could help me get the iterator part right? Sorry for the trouble 😅